### PR TITLE
Members-Only Project Pull Request

### PR DESCRIPTION
**Pull request by FdI96**
Kindly request a Pull Request.

We implemented the CRUD functionality.
Users can create a Post.
Usernames are displayed just for Users.
Users can only create a Post when logged in.
We implemented authentication using the devise gem.

- User can Sign In.
- User can Sign Up.
- User can Sign Out.

### CODE REVIEW

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️

Hello team, you have shown a great start for your project, keep up this energy.

## To Highlight
- You have a very clear PR description 👏
- Super great README
- Clear commit messages

Way to go 👏

## To Improve (MANDATORY)
So far so well, however, there are few defects that you need to fix in order to meet this project's requirements. Kindly improve your project based on these reviews below.

#### Behaviour
- [x] As a signed-in user, I should be able to see a post's author. Your app shows the author's ID and the signed-in user's username, please fix this ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/signed_in_can_see_author_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))
- [x] Your application encountered an error when I tried to sign out, user should be able to sign_out ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_out_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

#### Authentication and User setup
- [x] The User should have **many posts**, I see you tried to add this but please note that this association should be plural, e.g, `has_many :posts` - [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_associations_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication) but with extra hint - [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/associations_hint_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49722686/) )\_

#### Post routes
- [x] You are expected to create the application routes using `resources :posts, only: [:new, :create, :index]`, that way you only include routes and actions you need ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

#### Create action
- [x] Please make sure that the `user_id` for post is the one from the signed_in user ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_create_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))
```ruby
# the best way to do this
@post = current_user.posts.build(post_params)

#this could also work
@post = Post.new(post_params)
@post.user_id = current_user.id

# IMPORTANT: user_id cannot be passed in params (security)
```

## Nice to have (OPTIONAL)
Note that these are optional suggestions, although you are not obliged to implement them, I that you do so to improve your project.
- [ ] You could have this helper method to check if user is signed_in? can be handy here
```ruby
# must be a helper method to be used in views
helper_method :signed_in?
def signed_in?
  !current_user.nil?
end
```
- [ ] It would be nice to display Flash messages for unsuccessful actions

Have fun coding, see you at the top 🤓💻

> #### <br>You can ping me on Slack @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) to draw my attention in case I do not respond quickly on here. I'm always available and will be glad to help.<br><br>

File: app/views/posts/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    <div id="error_explanation">
+      <h2><%= pluralize(post.errors.count, "error") %> prohibited this post from being saved:</h2>
+
+      <ul>
+        <% post.errors.each do |error| %>
+          <li><%= error.full_message %></li>
+        <% end %>
+      </ul>
+    </div>
+  <% end %>
```
Comment (acushlakoncept) [inline app/views/posts/_form.html.erb]: - [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)

Use a helper method to display errors

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

> You've got a well written ReadMe file, keep it up 💯 👍 
> Nice use of helper methods 👏 

You've done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

- [x] Do not forget to add `yarn` to your setup instructions, perfect after the `bundle install`. I couldn't run the project without running `yarn`

- [x] Don't you think it's not proper to have the *Sign out* link visible to your app users when they are not yet signed in. Add a logic to only show that when they are signed in. Also the *Sign in and Sign up* link should not be visible when once they are signed in.

- [x] After creating a post, I encountered an error. 
Screen Shot 2021-04-03 at 6 19 02 AM  I noticed you do not even need that line of code `<%= post.user.user_id %>` Eliminating it fixed the issue.

- [x] Currently, I am able to create an empty posts, this should not be so, kindly add validations to your posts title and body.

- [x] [OPTIONAL] => Your commit messages should be capitalized, I see some of them are, just maintain consistency with the approach. You can watch this [video](https://www.loom.com/share/00771f369b7b47fc968c019ecf0d3ba6) on how to edit your commit messages using git rebase
 See [Git Commit Best Practices](https://github.com/trein/dev-best-practices/wiki/Git-Commit-Best-Practices#formatting-rules) for more insights

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by voscarmv — APPROVED** [review]
# Project status: Approved 💯 ✨ 😎 👍 

- Excellent work applying the necessary changes. Your app is now complete.

### You may merge this branch and move forward ⛙🛣️ 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻
# Happy Coding! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**